File: src/cli/decryptQuarantine.ts
import { readFile } from "node:fs/promises";
import process from "node:process";
import {
  decryptQuarantineEnvelope,
  parseEncryptedQuarantineEnvelope,
  type DecryptedQuarantineObject,
  type EncryptedQuarantineEnvelope,
} from "../quarantine/envelopeCrypto.js";

type InputStream = AsyncIterable<string | Uint8Array>;
type OutputStream = { write(value: string): unknown };

export interface DecryptQuarantineCliIo {
  readonly stdin: InputStream;
  readonly stdout: OutputStream;
  readonly stderr: OutputStream;
}

export function extractEnvelope(value: unknown): EncryptedQuarantineEnvelope {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    const object = value as Record<string, unknown>;
    if ("encrypted" in object) {
      return parseEncryptedQuarantineEnvelope(object.encrypted);
    }
  }
  return parseEncryptedQuarantineEnvelope(value);
}

export async function readPrivateKey(stdin: InputStream): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const chunk of stdin) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  }
  const value = Buffer.concat(chunks).toString("utf8").trim();
  if (!value) throw new Error("private key is required on stdin");
  return value;
}

export async function decryptQuarantineResponseFile(
  responsePath: string,
  privateKey: string,
): Promise<DecryptedQuarantineObject> {
  const response = JSON.parse(await readFile(responsePath, "utf8")) as unknown;
  return decryptQuarantineEnvelope(extractEnvelope(response), privateKey);
}

export async function runDecryptQuarantineCli(
  args: readonly string[],
  io: DecryptQuarantineCliIo,
): Promise<number> {
  try {
    const [responsePath] = args;
    if (!responsePath) {
      throw new Error(
        "usage: <private-key-command> | node dist/src/cli/decryptQuarantine.js <encrypted-response.json>",
      );
    }

    const privateKey = await readPrivateKey(io.stdin);
    const decrypted = await decryptQuarantineResponseFile(
      responsePath,
      privateKey,
    );
    const visible = escapedReviewValue(decrypted);
    if (visible.changed) {
      io.stderr.write(
        "warning: decrypted evidence contains invisible or control characters; stdout preserves the original evidence unchanged.\n",
      );
      io.stderr.write(
        `escaped visible representation:\n${JSON.stringify(visible.value, null, 2)}\n`,
      );
    }
    io.stdout.write(`${JSON.stringify(decrypted, null, 2)}\n`);
    return 0;
  } catch (error: unknown) {
    const message =
      error instanceof Error ? error.message : "decryption failed";
    io.stderr.write(`decrypt-quarantine failed: ${message}\n`);
    return 1;
  }
}

function escapedReviewValue(value: unknown): {
  changed: boolean;
  value: unknown;
} {
  let changed = false;
  const visit = (candidate: unknown): unknown => {
    if (typeof candidate === "string") {
      const escaped = escapeControls(candidate);
      if (escaped !== candidate) changed = true;
      return escaped;
    }
    if (Array.isArray(candidate)) return candidate.map(visit);
    if (candidate && typeof candidate === "object") {
      return Object.fromEntries(
        Object.entries(candidate).map(([key, child]) => {
          const escapedKey = escapeControls(key);
          if (escapedKey !== key) changed = true;
          return [escapedKey, visit(child)];
        }),
      );
    }
    return candidate;
  };
  const visible = visit(value);
  return { changed, value: visible };
}

function escapeControls(content: string): string {
  let escaped = "";
  for (const character of content) {
    const codePoint = character.codePointAt(0);
    if (!isInvisibleOrControl(codePoint)) {
      escaped += character;
    } else if (character === "\n") escaped += "\\n";
    else if (character === "\r") escaped += "\\r";
    else if (character === "\t") escaped += "\\t";
    else if (codePoint !== undefined) {
      escaped +=
        codePoint <= 0xffff
          ? `\\u${codePoint.toString(16).toUpperCase().padStart(4, "0")}`
          : `\\u{${codePoint.toString(16).toUpperCase()}}`;
    }
  }
  return escaped;
}

function isInvisibleOrControl(codePoint: number | undefined): boolean {
  return (
    codePoint !== undefined &&
    (codePoint <= 0x1f ||
      (codePoint >= 0x7f && codePoint <= 0x9f) ||
      codePoint === 0x200b ||
      codePoint === 0x200c ||
      codePoint === 0x200d ||
      codePoint === 0x2060 ||
      (codePoint >= 0xfe00 && codePoint <= 0xfe0f) ||
      (codePoint >= 0xe0000 && codePoint <= 0xe007f))
  );
}

if (import.meta.url === `file://${process.argv[1]}`) {
  runDecryptQuarantineCli(process.argv.slice(2), {
    stdin: process.stdin,
    stdout: process.stdout,
    stderr: process.stderr,
  }).then((exitCode) => {
    process.exitCode = exitCode;
  });
}

File: memory_router/cli/decrypt_quarantine.py
from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

from memory_router.envelope import decrypt_envelope


def extract_envelope(value: Any) -> Any:
    if isinstance(value, dict) and "encrypted" in value:
        return value["encrypted"]
    return value


def escaped_review_value(value: Any) -> tuple[bool, Any]:
    changed = False

    def visit(candidate: Any) -> Any:
        nonlocal changed
        if isinstance(candidate, str):
            escaped = _escape_string(candidate)
            changed = changed or escaped != candidate
            return escaped
        if isinstance(candidate, list):
            return [visit(item) for item in candidate]
        if isinstance(candidate, dict):
            result: dict[str, Any] = {}
            for key, child in candidate.items():
                escaped_key = _escape_string(str(key))
                changed = changed or escaped_key != key
                result[escaped_key] = visit(child)
            return result
        return candidate

    return changed, visit(value)


def _escape_string(value: str) -> str:
    result: list[str] = []
    for character in value:
        codepoint = ord(character)
        if not _is_invisible_or_control(codepoint):
            result.append(character)
        elif character == "\n":
            result.append("\\n")
        elif character == "\r":
            result.append("\\r")
        elif character == "\t":
            result.append("\\t")
        elif codepoint <= 0xFFFF:
            result.append(f"\\u{codepoint:04X}")
        else:
            result.append(f"\\u{{{codepoint:X}}}")
    return "".join(result)


def _is_invisible_or_control(codepoint: int) -> bool:
    return (
        codepoint <= 0x1F
        or 0x7F <= codepoint <= 0x9F
        or codepoint in {0x200B, 0x200C, 0x200D, 0x2060}
        or 0xFE00 <= codepoint <= 0xFE0F
        or 0xE0000 <= codepoint <= 0xE007F
    )


def run(path: str) -> int:
    try:
        key_text = sys.stdin.read().strip()
        if not key_text:
            raise ValueError("decryption key is required on stdin")
        value = json.loads(Path(path).read_text(encoding="utf-8"))
        decrypted = decrypt_envelope(extract_envelope(value), key_text)
        changed, visible = escaped_review_value(decrypted)
        if changed:
            print(
                "warning: decrypted evidence contains invisible or control characters; "
                "stdout preserves the original evidence unchanged.",
                file=sys.stderr,
            )
            print("escaped visible representation:", file=sys.stderr)
            print(json.dumps(visible, indent=2, ensure_ascii=False), file=sys.stderr)
        print(json.dumps(decrypted, indent=2, ensure_ascii=False))
        return 0
    except Exception as exc:
        print(f"decrypt-quarantine failed: {exc}", file=sys.stderr)
        return 1


def main() -> None:
    if len(sys.argv) != 2:
        print("usage: memory-router-decrypt-quarantine <encrypted-response.json>", file=sys.stderr)
        raise SystemExit(1)
    raise SystemExit(run(sys.argv[1]))


if __name__ == "__main__":
    main()
